File: week11/problem-set/prob2.py
"""Prob2 example.

This file belongs to the week 11 dynamic programming and pattern-matching
practice solutions. It computes the longest common subsequence of two strings.
"""


def lcs(x, y):
    """Return both the length and one example of the longest common subsequence.

    Args:
        x (int): The x input used by this function.
        y (int): The y input used by this function.

    Returns:
        tuple: The tuple of values produced by the algorithm.
    """
    m = len(x)
    n = len(y)
    table = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if x[i - 1] == y[j - 1]:
                table[i][j] = table[i - 1][j - 1] + 1
            else:
                table[i][j] = max(table[i - 1][j], table[i][j - 1])

    # Walk backwards through the table to rebuild one valid subsequence.
    i = m
    j = n
    characters = []

    while i > 0 and j > 0:
        if x[i - 1] == y[j - 1]:
            characters.append(x[i - 1])
            i -= 1
            j -= 1
        elif table[i - 1][j] >= table[i][j - 1]:
            i -= 1
        else:
            j -= 1

    return table[m][n], "".join(reversed(characters))


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    result = lcs("intention", "execution")
    print("LCS:", result)


if __name__ == "__main__":
    main()

File: week11/problem-set/prob2.cpp
/**
 * @file prob2.cpp
 * @brief C++ version of the longest-common-subsequence example.
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

using namespace std;

/**
 * @brief Return the length and one example of the longest common subsequence.
 * @param x const string& First input string.
 * @param y const string& Second input string.
 * @return pair<int, string> Pair containing the LCS length and one valid subsequence.
 */
pair<int, string> lcs(const string& x, const string& y) {
    int m = static_cast<int>(x.size());
    int n = static_cast<int>(y.size());
    vector<vector<int>> table(m + 1, vector<int>(n + 1, 0));

    for (int i = 1; i <= m; ++i) {
        for (int j = 1; j <= n; ++j) {
            if (x[i - 1] == y[j - 1]) {
                table[i][j] = table[i - 1][j - 1] + 1;
            } else {
                table[i][j] = max(table[i - 1][j], table[i][j - 1]);
            }
        }
    }

    int i = m;
    int j = n;
    string characters;
    while (i > 0 && j > 0) {
        if (x[i - 1] == y[j - 1]) {
            characters.push_back(x[i - 1]);
            --i;
            --j;
        } else if (table[i - 1][j] >= table[i][j - 1]) {
            --i;
        } else {
            --j;
        }
    }
    reverse(characters.begin(), characters.end());
    return {table[m][n], characters};
}

/**
 * @brief Run the LCS demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    pair<int, string> result = lcs("intention", "execution");
    cout << "LCS: (" << result.first << ", " << result.second << ")\n";
    return 0;
}
